File: docs/multicard_client.py
import os
import time
import hashlib
import logging
from datetime import datetime
from typing import Optional, Dict, Any

import requests

logger = logging.getLogger(__name__)

# =====================================================
# ENV CONFIG
# =====================================================
MULTICARD_BASE_URL = os.environ.get("MULTICARD_BASE_URL", "https://mesh.multicard.uz").rstrip("/")
MULTICARD_APPLICATION_ID = os.environ.get("MULTICARD_APPLICATION_ID", "").strip()
MULTICARD_SECRET = os.environ.get("MULTICARD_SECRET", "").strip()
MULTICARD_STORE_ID = int(os.environ.get("MULTICARD_STORE_ID", "0"))

AUTH_URL = f"{MULTICARD_BASE_URL}/auth"
PAYMENT_URL = f"{MULTICARD_BASE_URL}/payment"

if not MULTICARD_APPLICATION_ID or not MULTICARD_SECRET or not MULTICARD_STORE_ID:
    raise RuntimeError("MULTICARD env vars are not set correctly")

# =====================================================
# AUTH CACHE
# =====================================================
_auth_cache = {
    "token": None,
    "expires_at": 0.0,
}

def _parse_expired_at(value: Optional[str]) -> float:
    """
    Multicard sometimes returns expired_at like "YYYY-MM-DD HH:MM:SS".
    """
    try:
        return datetime.strptime(value, "%Y-%m-%d %H:%M:%S").timestamp()
    except Exception:
        return 0.0


def get_token(force: bool = False) -> str:
    now = time.time()

    # cached token still valid
    if (
        not force
        and _auth_cache["token"]
        and (_auth_cache["expires_at"] - 15) > now
    ):
        return _auth_cache["token"]

    resp = requests.post(
        AUTH_URL,
        json={
            "application_id": MULTICARD_APPLICATION_ID,
            "secret": MULTICARD_SECRET,
        },
        timeout=15,
    )

    if not resp.ok:
        logger.error("[Multicard] auth failed %s %s", resp.status_code, resp.text)
        resp.raise_for_status()

    data = resp.json()
    token = data.get("access_token") or data.get("token")
    expires_at = _parse_expired_at(data.get("expired_at"))

    if not token:
        raise RuntimeError(f"Invalid auth response: {data}")

    _auth_cache["token"] = token
    _auth_cache["expires_at"] = expires_at or (now + 600)

    logger.info("[Multicard] token refreshed, exp=%s", _auth_cache["expires_at"])
    return token


# =====================================================
# PAYMENT
# =====================================================
def create_payment(
    *,
    amount: int,                 # ❗ ТИЙИНЫ
    invoice_id: str,
    payment_system: str,
    return_url: str,
    callback_url: str,
    lang: str = "ru",
    billing_id: str | None = None,
) -> Dict[str, Any]:
    token = get_token()

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}",
    }

    payload: Dict[str, Any] = {
        "store_id": MULTICARD_STORE_ID,
        "amount": int(amount),
        "invoice_id": str(invoice_id),
        "payment_system": str(payment_system),
        "return_url": str(return_url),
        "callback_url": str(callback_url),
        "lang": str(lang),
        "details": {
            "invoice_id": str(invoice_id)
        },
    }

    if billing_id:
        payload["billing_id"] = str(billing_id)

    logger.info("[Multicard] create_payment payload=%s", payload)

    r = requests.post(PAYMENT_URL, json=payload, headers=headers, timeout=20)

    # if token expired mid-flight -> retry once
    if r.status_code == 401:
        logger.warning("[Multicard] 401 on create_payment, refreshing token and retrying once")
        token = get_token(force=True)
        headers["Authorization"] = f"Bearer {token}"
        r = requests.post(PAYMENT_URL, json=payload, headers=headers, timeout=20)

    if not r.ok:
        raise RuntimeError(f"Multicard error {r.status_code}: {r.text}")

    data = r.json()

    # Some APIs return {success:true,data:{...}}
    # If they return straight object - we keep it too.
    if isinstance(data, dict) and data.get("success") is False:
        raise RuntimeError(f"Multicard rejected payment: {data}")

    # normalize result
    if isinstance(data, dict) and "data" in data and data.get("success") is True:
        return data["data"]

    return data

def get_payment_info(payment_uuid: str) -> dict:
    """
    Запрашивает статус платежа по uuid через Multicard API.
    Используется как fallback, если sign не сошёлся.
    """
    if not payment_uuid:
        raise ValueError("payment_uuid is empty")

    token = get_token()

    # ✅ правильный URL
    url = f"{MULTICARD_BASE_URL}/payment/{payment_uuid}"

    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/json",
    }

    resp = requests.get(url, headers=headers, timeout=20)

    # если токен умер -> обновим и повторим 1 раз
    if resp.status_code == 401:
        token = get_token(force=True)
        headers["Authorization"] = f"Bearer {token}"
        resp = requests.get(url, headers=headers, timeout=20)

    resp.raise_for_status()
    return resp.json()


# =====================================================
# CALLBACK SIGN VERIFY (FIXED)
# =====================================================
def verify_callback_sign_payload(payload: dict, secret: str) -> bool:
    """
    Multicard callback sign check:
    1) берем payload без sign
    2) убираем пустые поля
    3) сортируем ключи
    4) key=value&... + secret
    5) md5
    """
    if not payload or not isinstance(payload, dict):
        return False

    got_sign = str(payload.get("sign") or "").strip().lower()
    if not got_sign:
        return False

    if not secret:
        return False

    # исключаем sign
    data = {k: payload.get(k) for k in payload.keys() if k != "sign"}

    # выкидываем None/"" поля
    cleaned = {}
    for k, v in data.items():
        if v is None:
            continue
        v_str = str(v).strip()
        if v_str == "":
            continue
        cleaned[k] = v_str

    # сортировка
    parts = [f"{k}={cleaned[k]}" for k in sorted(cleaned.keys())]

    base = "&".join(parts) + secret

    md5_calc = hashlib.md5(base.encode("utf-8")).hexdigest().lower()

    return md5_calc == got_sign

File: apps/api/src/modules/payments/multicard.client.ts
/**
 * Multicard payment gateway client.
 * Based on .cursor/docs/payment-reference/multicard_client.py
 */

import crypto from 'crypto';

const BASE_URL = (process.env.MULTICARD_BASE_URL ?? 'https://mesh.multicard.uz').replace(/\/$/, '');
const APPLICATION_ID = (process.env.MULTICARD_APPLICATION_ID ?? '').trim();
const SECRET = (process.env.MULTICARD_SECRET ?? '').trim();
const STORE_ID = parseInt(process.env.MULTICARD_STORE_ID ?? '0', 10);

const AUTH_URL = `${BASE_URL}/auth`;
const PAYMENT_URL = `${BASE_URL}/payment`;

let authCache: { token: string | null; expiresAt: number } = {
  token: null,
  expiresAt: 0,
};

function parseExpiredAt(value: string | null | undefined): number {
  if (!value) return 0;
  try {
    const date = new Date(value.replace(' ', 'T'));
    return date.getTime() / 1000;
  } catch {
    return 0;
  }
}

async function getToken(force = false): Promise<string> {
  const now = Date.now() / 1000;
  if (
    !force &&
    authCache.token &&
    authCache.expiresAt - 15 > now
  ) {
    return authCache.token;
  }

  const resp = await fetch(AUTH_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      application_id: APPLICATION_ID,
      secret: SECRET,
    }),
  });

  if (!resp.ok) {
    const text = await resp.text();
    console.error('[Multicard] auth failed', resp.status, text);
    throw new Error(`Multicard auth failed: ${resp.status}`);
  }

  const data = (await resp.json()) as Record<string, unknown>;
  const token = (data.access_token ?? data.token) as string | undefined;
  const expiresAt = parseExpiredAt(data.expired_at as string | undefined);

  if (!token) {
    throw new Error(`Invalid Multicard auth response: ${JSON.stringify(data)}`);
  }

  authCache = {
    token,
    expiresAt: expiresAt || now + 600,
  };
  return token;
}

export interface CreatePaymentParams {
  amount: number;       // tiyin
  invoiceId: string;
  paymentSystem: string;
  returnUrl: string;
  callbackUrl: string;
  lang?: string;
  billingId?: string;
}

export interface CreatePaymentResult {
  checkout_url?: string;
  check_url?: string;
  redirect_url?: string;
  payment_url?: string;
  url?: string;
  link?: string;
  uuid?: string;
  [key: string]: unknown;
}

/** Get redirect URL from Multicard payment response (different PS may use different keys). */
export function getPaymentRedirectUrl(payment: CreatePaymentResult): string | undefined {
  const raw =
    payment.checkout_url ??
    payment.check_url ??
    payment.redirect_url ??
    payment.payment_url ??
    payment.url ??
    payment.link;
  return typeof raw === 'string' && raw.trim() ? raw.trim() : undefined;
}

export async function createMulticardPayment(params: CreatePaymentParams): Promise<CreatePaymentResult> {
  const token = await getToken();
  const payload = {
    store_id: STORE_ID,
    amount: params.amount,
    invoice_id: params.invoiceId,
    payment_system: params.paymentSystem,
    return_url: params.returnUrl,
    callback_url: params.callbackUrl,
    lang: params.lang ?? 'ru',
    details: { invoice_id: params.invoiceId },
    ...(params.billingId ? { billing_id: params.billingId } : {}),
  };

  console.log('[Multicard] Sending payload:', JSON.stringify(payload, null, 2));

  let resp = await fetch(PAYMENT_URL, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify(payload),
  });

  if (resp.status === 401) {
    const newToken = await getToken(true);
    resp = await fetch(PAYMENT_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${newToken}`,
      },
      body: JSON.stringify(payload),
    });
  }

  if (!resp.ok) {
    const text = await resp.text();
    throw new Error(`Multicard payment error ${resp.status}: ${text}`);
  }

  const data = (await resp.json()) as Record<string, unknown>;
  if (typeof data.success === 'boolean' && data.success === false) {
    throw new Error(`Multicard rejected payment: ${JSON.stringify(data)}`);
  }
  if (data.success === true && data.data) {
    return data.data as CreatePaymentResult;
  }
  return data as CreatePaymentResult;
}

export async function getMulticardPaymentInfo(paymentUuid: string): Promise<Record<string, unknown>> {
  if (!paymentUuid) throw new Error('payment_uuid is empty');
  const token = await getToken();
  const url = `${BASE_URL}/payment/${paymentUuid}`;
  let resp = await fetch(url, {
    headers: { Authorization: `Bearer ${token}`, Accept: 'application/json' },
  });
  if (resp.status === 401) {
    const newToken = await getToken(true);
    resp = await fetch(url, {
      headers: { Authorization: `Bearer ${newToken}`, Accept: 'application/json' },
    });
  }
  if (!resp.ok) throw new Error(`Multicard get payment failed: ${resp.status}`);
  return (await resp.json()) as Record<string, unknown>;
}

/**
 * Verify callback sign: sorted key=value&... + secret, then MD5.
 * From multicard_client.py verify_callback_sign_payload.
 */
export function verifyCallbackSignPayload(payload: Record<string, unknown>, secret: string): boolean {
  const gotSign = String(payload?.sign ?? '').trim().toLowerCase();
  if (!gotSign || !secret) return false;

  const data: Record<string, string> = {};
  for (const k of Object.keys(payload)) {
    if (k === 'sign') continue;
    const v = payload[k];
    if (v == null) continue;
    const vStr = String(v).trim();
    if (vStr === '') continue;
    data[k] = vStr;
  }
  const parts = Object.keys(data)
    .sort()
    .map((k) => `${k}=${data[k]}`);
  const base = parts.join('&') + secret;
  const md5 = crypto.createHash('md5').update(base, 'utf8').digest('hex').toLowerCase();
  return md5 === gotSign;
}

/**
 * Alternative sign: fixed field order, concatenate values + secret, then MD5 or SHA1.
 * From guest_module.py multicard_callback.
 */
const CALLBACK_FIELDS_ORDER = [
  'store_id', 'amount', 'invoice_id', 'invoice_uuid', 'billing_id',
  'payment_time', 'phone', 'card_pan', 'card_token', 'ps', 'uuid', 'receipt_url',
];

export function verifyCallbackSignFixedOrder(payload: Record<string, unknown>, secret: string): boolean {
  const gotSign = String(payload?.sign ?? '').trim().toLowerCase();
  if (!gotSign || !secret) return false;

  const values = CALLBACK_FIELDS_ORDER.map((k) => {
    const v = payload[k];
    return v == null ? '' : String(v);
  });
  const base = values.join('') + secret;
  const md5 = crypto.createHash('md5').update(base, 'utf8').digest('hex').toLowerCase();
  const sha1 = crypto.createHash('sha1').update(base, 'utf8').digest('hex').toLowerCase();
  return gotSign === md5 || gotSign === sha1;
}

export function isMulticardConfigured(): boolean {
  return Boolean(APPLICATION_ID && SECRET && STORE_ID);
}
